cli/src/cmd/forge/script/build.rs

```diff
@@ -0,0 +1,189 @@
+use crate::{cmd::get_cached_entry_by_name, compile, opts::forge::ContractInfo};
+
+use ethers::{
+    prelude::{
+        artifacts::Libraries, cache::SolFilesCache, ArtifactId, Project, ProjectCompileOutput,
+    },
+    solc::artifacts::{CompactContractBytecode, ContractBytecode, ContractBytecodeSome},
+    types::{Address, U256},
+};
+
+use foundry_utils::PostLinkInput;
+use std::{collections::BTreeMap, str::FromStr};
+
+use super::*;
+
+impl ScriptArgs {
+    /// Compiles the file with auto-detection and compiler params.
+    pub fn build(&mut self, script_config: &ScriptConfig) -> eyre::Result<BuildOutput> {
+        let (project, output) = self.get_project_and_output(script_config)?;
+
+        let (contracts, sources) = output.into_artifacts_with_sources();
+        let contracts: BTreeMap<ArtifactId, CompactContractBytecode> =
+            contracts.into_iter().map(|(id, artifact)| (id, artifact.into())).collect();
+
+        let mut output = self.link(
+            project,
+            contracts,
+            script_config.config.parsed_libraries()?,
+            script_config.evm_opts.sender,
+            script_config.sender_nonce,
+        )?;
+        output.sources = sources.into_ids().collect();
+        Ok(output)
+    }
+
+    pub fn link(
+        &self,
+        project: Project,
+        contracts: BTreeMap<ArtifactId, CompactContractBytecode>,
+        libraries_addresses: Libraries,
+        sender: Address,
+        nonce: U256,
+    ) -> eyre::Result<BuildOutput> {
+        let mut run_dependencies = vec![];
+        let mut contract =
+            CompactContractBytecode { abi: None, bytecode: None, deployed_bytecode: None };
+        let mut highlevel_known_contracts = BTreeMap::new();
+
+        let mut target_fname = dunce::canonicalize(&self.path)
+            .expect("Couldn't convert contract path to absolute path")
+            .to_str()
+            .expect("Bad path to string")
+            .to_string();
+
+        let no_target_name = if let Some(target_name) = &self.target_contract {
+            target_fname = target_fname + ":" + target_name;
+            false
+        } else {
+            true
+        };
+
+        let mut extra_info = ExtraLinkingInfo {
+            no_target_name,
+            target_fname,
+            contract: &mut contract,
+            dependencies: &mut run_dependencies,
+            matched: false,
+            target_id: None,
+        };
+
+        // link_with_nonce_or_address expects absolute paths
+        let mut libs = libraries_addresses.clone();
+        for (file, libraries) in libraries_addresses.libs.iter() {
+            if file.is_relative() {
+                let mut absolute_path = project.root().clone();
+                absolute_path.push(file);
+                libs.libs.insert(absolute_path, libraries.clone());
+            }
+        }
+
+        foundry_utils::link_with_nonce_or_address(
```

Comment:
this function is quite a lot -.-

let's keep it like it is but we definitely should consider refactoring this

Reply:
this was already like that from before unfortunately